app/src/main/java/ch/epfl/sweng/studdybuddy/firebase/Metabase.java

```diff
@@ -85,15 +96,79 @@ public void accept(List<Pair> pairs) {
     }
 
     public void putAllCourses(List<String> courseSelection, String userid) {
-        for(String course : courseSelection){
+        for (String course : courseSelection) {
             Pair pair = new Pair(userid, course);
             db.select("userCourse").select(Helper.hashCode(pair).toString()).setVal(pair);
         }
     }
+
     public void addListenner(AdapterAdapter ad) {
         this.ads.add(ad);
     }
 
-    public void clearListeners() { this.ads.clear(); }
+    public void clearListeners() {
+        this.ads.clear();
+    }
+
+
+    public ValueEventListener getBuddies(String uid, List<User> users) {
+        return getBuddiesAndConsume(uid, users, new LinkedList<>(), Consumer.doNothing());
+    }
+
+    public ValueEventListener getBuddiesAndConsume(String uid, List<User> users, List<String> uIdsToFill, Consumer<List<User>> consumer) {
+        return db.select(Messages.FirebaseNode.BUDDIES).getAll(Buddy.class, new Consumer<List<Buddy>>() {
+            @Override
+            public void accept(@Nullable List<Buddy> buddies) {
+                for (Buddy buddy : buddies) {
+                    String bob = buddy.buddyOf(uid);
+                    if (bob != null) {
+                        uIdsToFill.add(bob);
+                    }
+                }
+
+                getUsersFromIdsAndConsume(uIdsToFill, users, consumer);
+            }
+        });
+    }
+
+
+    public ValueEventListener fetchUser(List<User> usernames) {
+        return db.select(Messages.FirebaseNode.USERS).getAll(User.class, new Consumer<List<User>>() {
+            @Override
+            public void accept(@Nullable List<User> users) {
+                usernames.clear();
+                for (User user : users) {
+                    usernames.add(user);
+                }
+                notif();
+            }
+        });
+    }
 
+    public ValueEventListener fetchUserNames(List<String> usernames) {
+        return db.select(Messages.FirebaseNode.USERS).getAll(User.class, new Consumer<List<User>>() {
+            @Override
+            public void accept(@Nullable List<User> users) {
+                usernames.clear();
+                for (User user : users) {
+                    usernames.add(user.getName());
+                }
+                notif();
+            }
+        });
+    }
+
+    public View.OnClickListener updateFriendsOnDone(String uId, List<User> friendsSelection, MetaGroupAdmin mga, Intentable mother) {
+        return new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                //mga.putAllFriends(friendsSelection, uId );
+                for (User friend : friendsSelection) {
+                    Buddy buddy = new Buddy(uId, friend.getUserID().getId());
+                    db.select(Messages.FirebaseNode.BUDDIES).select(buddy.hash()).setVal(buddy);
+                    mother.launch();
+                }
```

Comment:
Arguably should be in a controller, not in the service part

Reply:
It could be in a controller, you are right.But so far this server has acceptable amount of functions. And I can extract this functions in another class (controller), but I suggest to do it after merge this branch and inviteFriends brunch in order to avoid a lot of conflicts.